### PR TITLE
set up package publishing

### LINKED ISSUE #363: Set up GitHub action workflow for publishing packages to NPM

### PR DESCRIPTION
**Pull request by Rugvip**
Based on https://github.com/Rugvip/lerna-publish-experiment.

Version bumps happen through PRs with commits created via local `yarn release`. Once merged into master, the workflows published all public packages to npm, and tags the commit with the version number.

Fixes #363

### CODE REVIEW

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    strategy:
+      matrix:
+        node-version: [12.x]
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: find location of yarn cache
+        id: yarn-cache
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+      - uses: actions/cache@v1
```
Comment (freben) [inline .github/workflows/master.yml]: give this a name too `cache yarn cache`

File: .github/workflows/master.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        with:
+          path: ${{ steps.yarn-cache.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
```
Comment (freben) [inline .github/workflows/master.yml]: i assume the glob is due to nohoist packages?

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        uses: actions/cache@v1
+        with:
+          path: node_modules
+          key: ${{ runner.os }}-modules-${{ hashFiles('yarn.lock') }}
+      - name: use node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v1
+        with:
+          node-version: ${{ matrix.node-version }}
+          registry-url: https://registry.npmjs.org/ # Needed for auth
+      - name: yarn install, build, and test
```
Comment (freben) [inline .github/workflows/master.yml]: `, lint`

File: .github/workflows/master.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      # Publishes current version of packages that are not already present in the registry
+      - name: publish
+        run: npx --no-install lerna publish from-package --yes
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+          CI: true
+      # Tags the commit with the version in the core package if the tag doesn't exist
+      - uses: Klemensas/action-autotag@1.2.3
+        with:
+          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
```
Comment (freben) [inline .github/workflows/master.yml]: i presume this is set automatically by the environment :)

File: packages/cli/templates/default-plugin/package.json.hbs [hunk 5]
```diff
@@ -4,7 +4,7 @@
   "main": "dist/cjs/index.js",
   "types": "dist/cjs/index.d.ts",
   "license": "Apache-2.0",
-  "private": false,
+  "private": true,
```
Comment (freben) [inline packages/cli/templates/default-plugin/package.json.hbs]: for now?

**Review by freben — COMMENTED** [review]
awesome, great work

File: packages/cli/templates/default-plugin/package.json.hbs (hunk 5 above)
Comment (Rugvip) [reply packages/cli/templates/default-plugin/package.json.hbs]: yeah for now :p

File: .github/workflows/master.yml (hunk 4 above)
Comment (Rugvip) [reply .github/workflows/master.yml]: Yup

File: .github/workflows/master.yml (hunk 1 above)
Comment (Rugvip) [reply .github/workflows/master.yml]: updated :all-the-things:

File: .github/workflows/master.yml (hunk 2 above)
Comment (Rugvip) [reply .github/workflows/master.yml]: Dunno tbh, I'm assuming this is a thing that's supposed to work in any repo and probably from when we had the `frontend/` folder. Not touching it here though :grin:

File: .github/workflows/master.yml (hunk 3 above)
Comment (Rugvip) [reply .github/workflows/master.yml]: Split them up like #383 instead, for the same reason

**Review by freben — APPROVED** [review]

**Comment by Rugvip** [comment]
Wanna merge #384 before this

**Comment by ewnd9** [comment]
Hi, what's the process for actually publishing packages to npm?

E.g. lighthouse plugin was [announced](https://backstage.io/blog/2020/04/06/lighthouse-plugin) at April but it's still has `private: true` in [`package.json`](https://github.com/spotify/backstage/blob/9d28108cb87ac060fd2d01f2dab955921ccaea7f/plugins/lighthouse/package.json#L8) and missing on npm - https://www.npmjs.com/package/@backstage/plugin-lighthouse

---

Not sure this is the best place to ask, but this pr is the most related thing I found via search

**Comment by Rugvip** [comment]
@ewnd9 Packages are published whenever there's a more up-to-date version to be published during a master build. So as soon as `private` is set to `false` in the plugin, it'll be published!

Wanna submit a PR to flip that switch? :grin:

**Comment by ewnd9** [comment]
Done :+1: https://github.com/spotify/backstage/pull/1272

**Pull request merged**